rfcs/0001-meta-schedule-autotensorir.md

```diff
@@ -0,0 +1,297 @@
+
+Method 2. Derive from `PyCompositeSchedule`, providing extra functionalities like initialization
+
+```python
+class MultiLevelTiling(PyCompositeSchedule):
+    def initialize(...):
+        ...
+    def apply(...):
+        ...
+```
+
+**Customize exploration strategies in python**. Developers can implement any search algorithm in python as well by deriving from `PySearchPolicy`.
+
+**Other customizable components**. This list includes:
+
+* Cost model
+* Database
+* Measure callbacks
+* Feature extractor
+* Program builder & runner
+* Analysis methods
+* ...
+
+In a short summary, almost every component of the system is decoupled with each other and extensions could be easily plugged in.
+
+### 4.4. Upstreaming Plan
+
+[M3a] Core infrastructure of the PPL
+* Instruction
+* Trace
+* Composite schedule
+* Sampler
+* Search policy
+* Design space generator
+
+[M3b] Host-side search infra
+* Database
+* Cost model
+* Measure callback
+
+[M3c] RPC-related search infra
+* Measure input, build result, measure result
+* Builder
+* Runner
+
+[M4a] Implementation of rules
+* Various built-in composite schedules
+* Various built-in mutators
+* Various built-in postprocessors
+* Automatic tensorization
+
+[M4b] Relay integration
+
+## 5. Drawbacks
+
+We are not aware of any drawbacks of the proposed system.
+
+## 6. Rationale and alternatives
+
+The system is designed with the principle of minimalism: different from alternative solutions, we do not require any change in existing codebase, or extra APIs to learn. It could potentially lower the bar of using automation systems. 
+
+Unifying manual scheduling, AutoTVM's semi automatic templates and AutoScheduler's (Ansor's) fully automatic sketch generation provides flexible way to balance injection new domain knowledge and automation.
+
+Flexibility in customization allows quick try-out on new tasks, new strategies and new hardware targets without deep knowledge of the system.
+
+## 7. Prior art
+
+**Tensor Expression (TE)** in TVM is a DSL that decouples compute and schedule, which provides convenient ways to handcraft optimized kernels for different hardware targets.
+
+**TensorIR** is the latest generation of TVM’s low-level IR. Its capability of eagerly applying schedule primitives opens the door for meta schedule, our proposed new-generation auto scheduling system.
+
+**AutoTVM** is the 1st generation automation framework in TVM, which requires developers to implement per-operator scheduling templates, and the system could handle the tuning process.
+
+**AutoScheduler (Ansor)** is the 2nd generation automation framework in TVM, whose built-in rules could automatically generate schedule templates for almost all the operators on CPU, GPU, etc.
+
+## 8. Unresolved questions
+
+**Supporting Control Flow and Assertions**
+
+Right now the meta schedule DSL does not support control flow. Although we didn’t see any real-world use case right now, it is possible that it could appear in some future workloads.
```

Comment:
The Autonomous Driving Algorithm has many control flow in the post process network. I want to know what is the definition and limitation of `control flow`? For example, if we could support `tf.cond` in relay completely, how will meta schedule to handle the next step?

Reply:
I think this refers to the scheduling language that transforms the program, not the program itself.